src/main/c/truffleposix/truffleposix.c

```diff
@@ -75,12 +76,42 @@ struct truffleposix_stat {
 static void copy_stat(struct stat *stat, struct truffleposix_stat* buffer);
 
 char* truffleposix_get_user_home(const char *name) {
-  struct passwd *entry = getpwnam(name);
-  if (entry != NULL) {
-    return entry->pw_dir;
-  } else {
+  struct passwd entry;
+  struct passwd *result = NULL;
+  int ret;
+
+  size_t buffer_size = sysconf(_SC_GETPW_R_SIZE_MAX);
+  if (buffer_size <= 0) {
+    buffer_size = 16384;
+  }
+
+  char *buffer = malloc(buffer_size);
+  if (buffer == NULL) {
     return NULL;
   }
+
+retry:
+  ret = getpwnam_r(name, &entry, buffer, buffer_size, &result);
+  if (result != NULL) {
+    char *home = strdup(entry.pw_dir);
```

Comment:
Since we know the upper-bound, I think calling `strndup(entry.pw_dir, buffer_size)` should be favored for safety.

Reply:
But this could still go past the buffer as the `pw_dir` string might be at an offset in `buffer`.
I could imagine some OS storing `pw_dir` directly in `struct passwd`, so then `buffer + buffer_size - entry.pw_dir` would make no sense and actually be harmful.
There is a guarantee `pw_dir` returns a NULL-terminated string, let's rely on that, like we did before.